File: FSharpImpureSegmenter/SegmentModule.fs
module SegmentModule

type Coordinate = (int * int) // x, y coordinate of a pixel
type Colour = byte list       // one entry for each colour band, typically: [red, green and blue]

type Segment = 
    | Pixel of Coordinate * Colour
    | Parent of Segment * Segment 



// HELPER
// Transpose a list of [r,g,b] values into a list of [ [r, ...], [g, ...], [b, ...] ]
let transpose (rows : list<list<'T>>) : list<list<'T>> =
    let n = List.length (List.head rows)
    List.init n (fun i -> (List.map (List.item i) rows))

// HELPER
// Flatten a segment into a list where each element is a segments colour
let rec desegment (segment:Segment) : list<list<float>> =
    match segment with
    | Pixel (_, colour) -> [List.map float colour]                              // The colour band of the pixel, as float values
    | Parent (seg01, seg02) -> (desegment seg01) @ (desegment seg02)            // Concatenate the two de-segmented segments

// return a list of the standard deviations of the pixel colours in the given segment
// the list contains one entry for each colour band, typically: [red, green and blue]
let stddev (segment: Segment) : float list =
    match segment with
    | Pixel (_, colour) -> List.map (fun _ -> 0.0) colour
    | _ ->
        desegment segment                                                       // First, flatten our segment tree into their colours ( Segment -> [ [r,g,b], ... ] )
        |> transpose                                                            // Now transpose into a list for each colour band ( [ [r,g,b], ... ] -> [ [r, ...], [g, ...], [b, ...] ] )
        |> List.map (fun colourN ->                                             // Now we can do the maths for each band...
            let len = List.length colourN
            let avg = List.average colourN
            List.map (fun colourValue -> pown (colourValue - avg) 2) colourN
            |> List.sum
            |> (*) (1.0 / float len)                                            // Really cool that operators are also functions
            |> sqrt)





// HELPER
// Calculate the number of pixels in a segment
let rec segmentSize (segment:Segment) : float =
    match segment with
    | Pixel (_, _) -> 1.0
    | Parent (seg1, seg2) -> (segmentSize seg1) + (segmentSize seg2)

// determine the cost of merging the given segments: 
// equal to the standard deviation of the combined the segments minus the sum of the standard deviations of the individual segments, 
// weighted by their respective sizes and summed over all colour bands
let mergeCost segment1 segment2 : float = 
    let combined = Parent (segment1, segment2)
    
    let combSize = segmentSize combined
    let seg1Size = segmentSize segment1
    let seg2Size = segmentSize segment2

    (stddev combined, stddev segment1, stddev segment2)                         // Calculate our standard deviations...
    |||> List.map3 (fun combStddevN seg1StddevN seg2StddevN ->                  // Now do the maths to calculate the cost for each band...
        combSize * combStddevN
        |> (+) ( - ((seg1StddevN * seg1Size) + (seg2StddevN * seg2Size))))
    |> List.sum

File: CSharpSegmenter/SegmentModule.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;

namespace CSharpSegmenter {

    interface Segment {

        Segment ParentSeg { get; set; }

    }

    class Pixel : Segment {

        private Segment parent;
        public Segment ParentSeg {
            get { return parent; }
            set { parent = value; }
        }

        public Tuple<int, int> Coordinate;
        public float[] Colour;

        public Pixel(Tuple<int, int> coordinate, byte[] colour) {
            this.Coordinate = coordinate;

            this.Colour = new float[colour.Length];

            for (int i = 0; i < colour.Length; i++)
                this.Colour[i] = colour[i];
        }

    }

    class Parent : Segment {

        private Segment parent;
        public Segment ParentSeg {
            get { return parent; }
            set { parent = value; }
        }

        public Segment Seg1;
        public Segment Seg2;

        public Parent(Segment seg1, Segment seg2) {
            this.Seg1 = seg1;
            this.Seg2 = seg2;

            seg1.ParentSeg = this;
            seg2.ParentSeg = this;
        }

    }


    class SegmentModule {

        public static int SegmentSize (Segment segment) {

            if (segment == null) return 0;
            else if (segment is Pixel) return 1;
            else {

                Parent seg = (Parent)segment;

                return SegmentSize(seg.Seg1) + SegmentSize(seg.Seg2);

            }

        }

        public static float[][] Transpose(float[][] segColours) {

            float[][] result = new float[segColours[0].Length][];

            for (int i = 0; i < segColours[0].Length; i++) result[i] = new float[segColours.Length];

            for (int i = 0; i < segColours.Length; i++)
                for (int j = 0; j < segColours[i].Length; j++)
                    result[j][i] = segColours[i][j];

            return result;

        }

        public static float[][] Desegment (Segment segment) {

            if (segment is Pixel) {

                float[][] result = { ((Pixel)segment).Colour };

                return result;

            } else {

                Parent seg = (Parent)segment;

                float[][] seg1Result = Desegment(seg.Seg1);
                float[][] seg2Result = Desegment(seg.Seg2);

                float[][] result = new float[seg1Result.Length + seg2Result.Length][];

                for (int i = 0; i < result.Length; i++) {

                    if (i <= seg1Result.Length - 1) result[i] = seg1Result[i];
                    else result[i] = seg2Result[i - seg1Result.Length];

                }

                return result;

            }

        }

        public static float[] Stddev (Segment segment) {

            if (segment is Pixel) {

                Pixel seg = (Pixel)segment;

                float[] result = new float[seg.Colour.Length];

                for (int i = 0; i < result.Length; i++) result[i] = 0F;

                return result;

            } else {

                float[][] colourBands = Transpose(Desegment(segment));
                float[] result = new float[colourBands.Length];

                for (int i = 0; i < colourBands.Length; i++) {
                    float len = colourBands[i].Length;
                    float avg = colourBands[i].Average();
                    float sum = 0F;

                    for (int j = 0; j < colourBands[i].Length; j++)
                        sum += (float)Math.Pow(colourBands[i][j] - avg, 2);

                    result[i] = (float)Math.Sqrt(sum / len);

                }

                return result;

            }

        }


        public static float MergeCost (Segment seg1, Segment seg2) {

            Parent combined = new Parent(seg1, seg2);

            int combSize = SegmentSize(combined);
            int seg1Size = SegmentSize(seg1);
            int seg2Size = SegmentSize(seg2);

            float[] combStddev = Stddev(combined);
            float[] seg1Stddev = Stddev(seg1);
            float[] seg2Stddev = Stddev(seg2);

            float sum = 0F;

            for (int i = 0; i < combStddev.Length; i++)
                sum += (combSize * combStddev[i]) - ((seg1Size * seg1Stddev[i]) + (seg2Size * seg2Stddev[i]));

            return sum;

        }

    }
}
